pkg/engine/jmespath/functions.go

```diff
@@ -1061,48 +1061,62 @@ func jpRandom(arguments []interface{}) (interface{}, error) {
 
 func jpX509Decode(arguments []interface{}) (interface{}, error) {
 	res := make(map[string]interface{})
+	var cert *x509.Certificate
+	var cr *x509.CertificateRequest
 	input, err := validateArg(x509_decode, arguments, 0, reflect.String)
 	if err != nil {
 		return nil, err
 	}
 	p, _ := pem.Decode([]byte(input.String()))
 	if p == nil {
-		return res, errors.New("invalid certificate")
+		return res, nil
 	}
 
-	cert, err := x509.ParseCertificate(p.Bytes)
+	cert, err = x509.ParseCertificate(p.Bytes)
 	if err != nil {
-		return res, err
-	}
-
-	buf := new(bytes.Buffer)
-	if fmt.Sprint(cert.PublicKeyAlgorithm) == "RSA" {
-		spki := cryptobyte.String(cert.RawSubjectPublicKeyInfo)
-		if !spki.ReadASN1(&spki, cryptobyte_asn1.SEQUENCE) {
-			return res, errors.New("writing asn.1 element to 'spki' failed")
-		}
-		var pkAISeq cryptobyte.String
-		if !spki.ReadASN1(&pkAISeq, cryptobyte_asn1.SEQUENCE) {
-			return res, errors.New("writing asn.1 element to 'pkAISeq' failed")
-		}
-		var spk asn1.BitString
-		if !spki.ReadASN1BitString(&spk) {
-			return res, errors.New("writing asn.1 bit string to 'spk' failed")
-		}
-		kk, err := x509.ParsePKCS1PublicKey(spk.Bytes)
+		cr, err = x509.ParseCertificateRequest(p.Bytes)
```

Comment:
Isn't there a way to determine what kind of object it is instead of trying cert first and fallback to csr if it fails ?

Reply:
It would be better if we just separate these two

Reply:
I'm not sure we need different filters but there's room for improvement here.

Reply:
Structure of the code should look something like this:
```
	if input, err := validateArg(x509_decode, arguments, 0, reflect.String); err != nil {
		return nil, err
	} else if block, _ := pem.Decode([]byte(input.String())); p == nil {
		return nil, errors.New("failed to decode PEM block")
	} else {
		switch block.Type{
		case "CERTIFICATE":
			// parse certificate
		case "CERTIFICATE REQUEST":
			// parse certificate request
		default:
			return nil, errors.New("PEM block neither contains a CERTIFICATE or CERTIFICATE REQUEST")
		}
		// rest of the parsing...
	}
```

Reply:
I refactored it, test cases and policies are passing and `x509_decode('xyz')` is returning an error

Reply:
Ultimately it could be as simple as this:
```
	if input, err := validateArg(x509_decode, arguments, 0, reflect.String); err != nil {
		return nil, err
	} else if block, _ := pem.Decode([]byte(input.String())); p == nil {
		return nil, errors.New("failed to decode PEM block")
	} else {
		switch block.Type {
		case "CERTIFICATE":
			if cert, err := x509.ParseCertificate(block.Bytes); err != nil {
				return nil, err
			} else if cert.PublicKeyAlgorithm == x509.RSA {
				return nil, errors.New("public key algorithm is not RSA")
			} else if pk, err := parseSubjectPublicKeyInfo(cert.RawSubjectPublicKeyInfo); err != nil {
				return nil, errors.New("failed to parse subject public key info")
			} else {
				csr.PublicKey = pk
				return encode(csr)
			}
		case "CERTIFICATE REQUEST":
			if csr, err := x509.ParseCertificateRequest(block.Bytes); err != nil {
				return nil, err
			} else if csr.PublicKeyAlgorithm == x509.RSA {
				return nil, errors.New("public key algorithm is not RSA")
			} else if pk, err := parseSubjectPublicKeyInfo(csr.RawSubjectPublicKeyInfo); err != nil {
				return nil, errors.New("failed to parse subject public key info")
			} else {
				csr.PublicKey = pk
				return encode(csr)
			}
		default:
			return nil, errors.New("PEM block neither contains a CERTIFICATE or CERTIFICATE REQUEST")
		}
	}
```